lib/core/include/replica.hpp

```diff
@@ -0,0 +1,312 @@
+            }
+        } // throw_if_replica_number_is_invalid
+
+        /// \brief Enforce valid path and replica number for replica operation
+        /// \param[in] _comm connection object
+        /// \param[in] _p path to inspect
+        /// \param[in] _rn replica number to inspect
+        /// \throws filesystem_error if the path is empty or too long
+        /// \throws irods::exception if the path does not refer to a data object or replica number is invalid
+        /// \since 4.2.9
+        template<typename rxComm>
+        inline auto throw_if_replica_input_is_invalid(
+            rxComm& _comm,
+            const fs::path& _p,
+            const replica_number_type _rn) -> void
+        {
+            fs::detail::throw_if_path_is_empty(_p);
+
+            fs::detail::throw_if_path_length_exceeds_limit(_p);
+
+            throw_if_path_is_not_a_data_object(_comm, _p);
+
+            throw_if_replica_number_is_invalid(_rn);
+        } // throw_if_replica_input_is_invalid
+
+    } // namespace detail
+
+    namespace
+    {
+        // TODO: make this into an API plugin
+        template<typename rxComm>
+        auto get_replica_info(
+            rxComm& _comm,
+            const fs::path& _p,
+            const replica_number_type _rn) -> std::vector<std::string>
+        {
+            query_builder qb;
+            qb.zone_hint(*fs::get_zone_name(_p));
+
+            const std::string qstr = fmt::format(
+                "SELECT "
+                "DATA_ID, "
+                "DATA_COLL_ID, "
+                "DATA_NAME, "
+                "DATA_REPL_NUM, "
+                "DATA_VERSION,"
+                "DATA_TYPE_NAME, "
+                "DATA_SIZE, "
+                "DATA_RESC_NAME, "
+                "DATA_PATH, "
+                "DATA_OWNER_NAME, "
+                "DATA_OWNER_ZONE, "
+                "DATA_REPL_STATUS, "
+                "DATA_STATUS, "
+                "DATA_CHECKSUM, "
+                "DATA_EXPIRY, "
+                "DATA_MAP_ID, "
+                "DATA_COMMENTS, "
+                "DATA_CREATE_TIME, "
+                "DATA_MODIFY_TIME, "
+                "DATA_MODE, "
+                "DATA_RESC_HIER, "
+                "DATA_RESC_ID "
+                "WHERE DATA_NAME = '{}' AND COLL_NAME = '{}' AND DATA_REPL_NUM = '{}'",
+                _p.object_name().string(), _p.parent_path().string(), _rn);
+
+            const auto q = qb.build<rxComm>(_comm, qstr);
+            if (q.size() <= 0) {
+                THROW(CAT_NO_ROWS_FOUND, "no replica information found");
+            }
+            return q.front();
+        } // get_replica_info
+
+        template<typename rxComm>
+        auto replica_mtime(
+            rxComm& _comm,
+            const fs::path& _p,
+            const replica_number_type _rn) -> std::string
+        {
+            const auto result = get_replica_info(_comm, _p, _rn);
+            return result[18]; // 18 == index of MODIFY_TIME column from SELECT statement
```

Comment:
yes please

Reply:
I can look into this. In the other branch I mentioned, I am constructing a map of the result set using the genquery column names as keys so that they can be referred to directly, regardless of ordering. Thinking on it further, I think having the indexes laid out in a SPOT may just be better/easier to manage overall anyway.